File: singly_linear_linked_list.py
#! /bin/python3
# This is singly linear linked list in Python
# First make Node class, It is used to store node data and connection to next node
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __str__(self):
        return "This is Node."

# Make linked list class
class SLLL:
    def __init__(self):
        self.first = None

    # Insertion into Linked list at specific position
    def insert(self, position, node):
        p = self.first
        while(position > 2):
            p = p.next
            position -= 1

        node.next = p.next
        p.next = node

    # Deletion at specific position
    def delete(self, position):
        p = self.first
        while position > 2:
            p = p.next
            position -= 1

        node = p.next
        p.next = p.next.next
        return node

    # Disply Linked list
    def display(self):
        p = self.first
        while True:
            print(p.data)
            p = p.next
            if p.next == None:
                print(p.data)
                break

    def __str__(self):
        return "This is Singly Linear Linked List."

# Start making and linking lists
# Create Nodes
list1 = Node("First Node")
list2 = Node("Second Node")
list3 = Node("Third Node")
list4 = Node("Fourth Node")
list5 = Node("Fifth Node")

# Here, linked_list is the object of SLLL. It must point first list to start linking
linked_list = SLLL()
linked_list.first = list1

# Linking all nodes
list1.next = list2
list2.next = list3
list3.next = list4
list4.next = list5

temp = Node("Inserted node")

print("Before Insertion:")
linked_list.display()
print("\nAfter Insertion:")
linked_list.insert(4, temp)
linked_list.display()

node = linked_list.delete(2)
print("Deleted Node: {}".format(node.data))

print("\nAfter deletion:")
linked_list.display()

File: singly_linear_linked_list.cpp
#include<iostream>
using std::cin;
using std::cout;
using std::endl;

// Singly linear linked list; Linked list is the connection of Nodes.
class SLLL{

    // Create node for the linked list
    class Node{
        public:
        int data;
        Node *next;
        Node(){
            next = NULL;
        }
    };
    
    Node *start;

    public:
    void display();

    void push(int data=0);
    void insert(int data);
    void insert(int data, int postion);

    void pop();
    void remove();
    void remove(int position);
};

// Remove specific node between first and last
void SLLL::remove(int position){
    Node *temp = new Node();
    Node *p = new Node();
    p = start;
    while(position > 2){
        p = p->next;
        position--;
    }
    temp = p->next;
    p->next = temp->next;
    delete temp;
    cout<<"Removed!!"<<endl;
}

// Remove first node
void SLLL::remove(){
    Node *temp = new Node();
    temp = start;
    start = start->next;
    delete temp;
    cout<<"Rmoved!!"<<endl;
}

// Remove last node
void SLLL::pop(){
    Node *temp = new Node();
    Node *p = new Node();
    p = start;
    while(p->next->next != NULL){
        p = p->next;
    }
    temp = p->next;
    p->next = NULL;
    delete temp;
    cout<<"Popped!!"<<endl;
}

// Insert new node at given position
void SLLL::insert(int data, int position){
    Node *temp = new Node();
    temp->data = data;
    Node *p = start;

    while(position > 2){
        p = p->next;
        position--;
    }
    temp->next = p->next;
    p->next = temp;
    cout<<"Inserted!!"<<endl;
}

// Insert new node at first
void SLLL::insert(int data){
    Node *temp = new Node();
    temp->data = data;
    temp->next = start;
    start = temp;
    cout<<"Inserted!!"<<endl;
}

// Insert New Node at Last
void SLLL::push(int data){
    Node *temp = new Node();
    temp->data = data;
    
    Node *p = start;
    while(p->next != NULL){
        p = p->next;
    }
    p->next = temp;
    cout<<"Pushed!!"<<endl;
}

// Display Linked List
void SLLL::display(){
    Node *p = start;
    while(p->next != NULL){
        cout<<p->data<<endl;
        p = p->next;
    }
    cout<<p->data<<endl;
}


int main(){

    SLLL *linked_list = new SLLL();
    linked_list->insert(5);
    linked_list->insert(6);
    linked_list->push(8);
    linked_list->insert(10, 3);
    linked_list->insert(11, 2);
    linked_list->push();

    cout<<"After Insertion:"<<endl;
    linked_list->display();

    linked_list->remove(3);
    linked_list->pop();
    linked_list->remove();
    cout<<"After Removed:"<<endl;
    linked_list->display();

    delete linked_list;
    return 0;
}